### ISSUE TITLE
CAPM3 (capm3-system namespace pods) should not be running among workloads

### ISSUE THREAD

**Issue by tuminoid** [metal3-io/cluster-api-provider-metal3#940]
**What steps did you take and what happened:**
As discussed in metal3-sec group related to https://github.com/metal3-io/baremetal-operator/pull/1241 CAPM3/IPAM (capm3-system namespace) should not be running among user workloads, but with k8s controlplane or infra nodes.

**What did you expect to happen:**
CAPM3/IPAM (capm3-system) is running on controlplane or in infra nodes, not among regular workload. These should be sensible default value, and documentation should be available to explain the options.

**Environment:**

- CAPM3 version: any/latest
- Environment (metal3-dev-env or other): any

/kind bug
/kind documentation
/assign @lentzi90

**Comment by Rozzii** [comment]
Would be welcomed I will move this to frozen, no need to always remove the stale
/lifecycle frozen

**Comment by fmuyassarov** [comment]
I think this request is trying to lock the capm3 controller to control plane nodes, filtered via well-known labels like `node-role.kubernetes.io/control-plane`. While I understand the intention of keeping the controller pods on control plane nodes, I believe this kind of restriction should be handled on the consumer side, not here.

Environments differ, and an affinity rule that works for one setup may not work for another. In my opinion, we should provide the bare minimum YAML that works across environments and let consumers/admins decide where to place the pods. In other words, admins should define their own affinity rules and patch the manifests before deployment.

Unless there’s a strong reason to enforce such restrictions by default, I think we should aim for provider/environment-agnostic manifests.

**Comment by fmuyassarov** [comment]
Perhaps as a recommendation, it could be explain in the documentation that this is a best practice and etc.

**Comment by tuminoid** [comment]
We should provide secure defaults as well. IMO it is better to provide those when they work for most people, and if someone is running non-compatible setups, they can customize per setup. Leaving this out means everybody needs to do that to be secure.

That said, maybe we should put this into separate kustomization as compromise?

**Comment by fmuyassarov** [comment]
Yes, I think you could have clean default + a kustomization that patches those lines. This should work.

**Comment by fmuyassarov** [comment]
And to be clear, originally I was against adding infra label that doesn't exist in the Kubernetes documentation.

**Comment by lentzi90** [comment]
I'll also add some of my thoughts from an offline chat earlier.
To me it is important provide a smooth first experience for the most simple use-cases. Users must be able to deploy with the defaults we provide in "most cases". This means following CAPI guides to do a simple `clusterctl init` and get something that actually runs without making any extra modifications.

I also think that secure defaults are important and following best practices.

I would absolutely not want to force scheduling on control-plane nodes by default. Users may want to dedicate control-plane for only control-plane controllers. Therefore I thought it was better to have an alternative: schedule on control-plane **or** infra nodes. My thought was that `infra` is then something the user has control over. They can then label the nodes they want and put taints, etc to keep CAPM3/IPAM isolated/grouped the way they want.

Regarding the infra node-role label I do not agree with @fmuyassarov . It is part of the well-known prefix and makes sense to me at least. It is generic enough and fills a common use-case (i.e. run "infrastructure" controllers grouped together, isolated from the rest). The point is not that it is commonly used, it is that the users can easily start using this if they see the need. It provides an escape hatch for not running on control-plane nodes.

I am also ok with not having any affinity or selectors by default. This is what most projects do as far as I can tell. For any kind of production setups, I expect the admins will be able to adjust the manifests as wanted, e.g. adding resources requests and limits, node selectors etc. For casual, getting-started users we then keep things as simple as they can be.
For reference, CAPI has good docs already on how to set node affinity using [CAPI operator](https://cluster-api-operator.sigs.k8s.io/topics/configuration/provider-spec-configuration). This is what I would expect production setups use.

If we decide to not set any affinity, I think we should also consider removing the tolerations. We do not have a *need* to run on control-plane nodes (unlike e.g. cloud-provider integrations that have to be able to run on the very first node), so the user should get a say in if we can run there or not.

Honestly, I am leaning towards simplicity. No affinity or tolerations at all and leave it up to advanced users to add as needed.

**Comment by fmuyassarov** [comment]
If `node-role.kubernetes.io/infra` was a commonly used label, or something typically set by installers like kubeadm, I’d have no issue allowing it in the default manifests. But since setting affinity against `node-role.kubernetes.io/infra` is based on assumptions (someone might have it) rather than common practice, I don’t find it a convincing justification.

**Comment by lentzi90** [comment]
> But since setting affinity against node-role.kubernetes.io/infra is based on assumptions (someone might have it) rather than common practice, I don’t find it a convincing justification.

It is the other way around. We don't assume that users have nodes labeled like this. The assumption is that they can *choose* to use this label on their nodes. So if they are not comfortable with running CAPM3 on the control-plane, then they can easily label a group of nodes with this label. CAPM3 would then be scheduled on those nodes instead with higher priority.

Why exactly this label? Because it is in the well-known prefix that is meant for things like this. Specifically `infra` fits well with CAPM3 in my opinion, since we are an infrastructure provider.

**Comment by fmuyassarov** [comment]
> > But since setting affinity against node-role.kubernetes.io/infra is based on assumptions (someone might have it) rather than common practice, I don’t find it a convincing justification.
> 
> So if they are not comfortable with running CAPM3 on the control-plane, then they can easily label a group of nodes with this label. 

Well, with this approach, some users will effectively be forced to do extra work — either by removing the affinity or by labeling their nodes just to be able to run CAPM3. So to me, this really comes down to who is responsible for setting labels/affinity.

**Comment by lentzi90** [comment]
Yes, as mentioned, I am leaning towards not setting any of this. But if we do, I think it is much better to have this extra label than to lock everything on the control-plane by default.

Thinking about writing tutorials for this, doing a `kubectl label` + `clusterctl init` is easier than `clusterctl generate provider > file.yaml` then edit the file and finally `kubectl apply`.

**Comment by fmuyassarov** [comment]
Yes, either none (preferably) or everything else makes sense. But I think it would be good to have a kustomization for those who want those affinities in place out of the box.

**Comment by fmuyassarov** [comment]
And I think a Helm based installation would fit really well here, since it allows users to easily pass these settings at install time. However, given the nature of the project especially the dependency on capi it’s a bit harder to encourage users to install it via Helm. For BMO though, this approach could make a lot more sense.

**Comment by lentzi90** [comment]
CAPI operator already has a nice [API to handle these](https://cluster-api-operator.sigs.k8s.io/topics/configuration/provider-spec-configuration). Kustomize can also already be used since we provide a base.

> But I think it would be good to have a kustomization for those who want those affinities in place out of the box.

Which ones do you think would make sense to have kustomizations for? From my perspective, it is not worth much to have specific kustomizations, since each user likely want their own thing anyway. As long as we provide the base, they can easily build on top.

I also agree that helm doesn't fit in the ecosystem (and I dislike i).

**Comment by fmuyassarov** [comment]
> CAPI operator already has a nice [API to handle these](https://cluster-api-operator.sigs.k8s.io/topics/configuration/provider-spec-configuration). Kustomize can also already be used since we provide a base.
> 
> > But I think it would be good to have a kustomization for those who want those affinities in place out of the box.
> 
> Which ones do you think would make sense to have kustomizations for? 


I was thinking about control-plane and infra but not because I really want it but rather due to @tuminoid's comment above. In my opinion, leaving those out of the tree is the best.

**Comment by fmuyassarov** [comment]
If we all agree that we better not add those labels and keep our manifests clean, then perhaps we could have mentioned it somewhere in the docs that users should set those affinities or labels on their own.

**Comment by vishalanarase** [comment]
Have we reached a final decision regarding the approach? This will help me align my changes correctly.

Thanks for any clarification!

**Comment by lentzi90** [comment]
There hasn't really been any more discussions 🙁 
I asked for more input on slack now, here: https://kubernetes.slack.com/archives/CHD49TLE7/p1771845729140539 
Based on the discussion here on the issue, I would say we should remove the tolerations and not add any affinity.

**Comment by tuminoid** [comment]
For CAPI ecosystem alignment, we should actually do nothing. CAPI and the providers all have tolerations, and (with quick check) only CAPA has soft affinity.  While I'd personally love to have soft affinity to prefer masters/cp, I can live with tolerations for ecosystem alignment and just document the hardening options including affinities (the latter part of the original issue).
